bpf/lib/nat.h

```diff
@@ -1042,31 +1041,25 @@ snat_v4_rev_nat_handle_icmp_frag_needed(struct __ctx_buff *ctx, __u64 off)
 	default:
 		return NAT_PUNT_TO_STACK;
 	}
-	state = snat_v4_lookup(&tuple);
-	if (!state)
+
+	*state = snat_v4_lookup(&tuple);
+	if (!*state)
 		return NAT_PUNT_TO_STACK;
 
 	/* We found SNAT entry to rev-NAT embedded packet. The source addr
 	 * should point to endpoint that initiated the packet, as-well if
 	 * dest port had been NATed.
 	 */
-	ret = snat_v4_icmp_rewrite_ingress_embedded(ctx, &tuple, state,
-						    off, icmpoff);
-	if (IS_ERR(ret))
-		return ret;
-
-	/* Switch back to the outer header. */
-	tuple.nexthdr = IPPROTO_ICMP;
```

Comment:
Yep, it assume that the outer headers match the inner headers.

Which isn't terribly robust - it would be a nice improvement to explicitly check whether we *can* rewrite the outer header with the inner packet's SNAT entry  (and drop otherwise)